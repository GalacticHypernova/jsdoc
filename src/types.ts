export declare const files: Map<string, string>;
/**
* A utility for extracting the JSDoc annotations from a member in the given module, if it exists.
  @param modulePath - The path to the module containing the function.
  @param member - The name of the member to extract the JSDoc annotations from.
  @returns The JSDoc annotations for the given function, if it exists.
  @throws {Error} if something went wrong
*/
export declare function extractJSDoc(modulePath: string, member: string): RegExpMatchArray | null | undefined;
