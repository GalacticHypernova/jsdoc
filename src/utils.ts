import { readFileSync, existsSync, statSync } from "node:fs"
import { resolve } from 'pathe'
export const files = new Map<string, string>();

/**
* A utility for extracting the JSDoc annotations from a function in the given module, if it exists.
  @param modulePath - The path to the module containing the function.
  @param functionName - The name of the function to extract the JSDoc annotations from.
  @returns The JSDoc annotations for the given function, if it exists.
  @throws {Error} if something went wrong
*/
export function extractJSDoc(modulePath: string, functionName: string) {
  try {
    const jsDocRE = new RegExp(`(\\/\\*\\*[\\s\\S]*\\*\\/)\\s*(?:\n[^\\n]*${functionName})`,'i')
    modulePath = resolve(modulePath.replaceAll('../',""))
    if (!files.has(modulePath)) {
      if (existsSync(modulePath + "/package.json")) {
        const pkg = JSON.parse(readFileSync(modulePath + "/package.json", "utf8"))
        files.set(modulePath, readFileSync(modulePath + "/" + pkg.main, "utf8"))
      }
      else {
        if(statSync(modulePath).isDirectory()) { modulePath = modulePath+"/index" }
        console.log(modulePath)
        for (const ext of [".ts",".js",".mjs",".cjs"]) {
          console.log("hi")
          if (existsSync(modulePath + ext)) {
            files.set(modulePath, readFileSync(modulePath+ext, "utf8"))
            break
          }
        }
      }
    }
    console.log(files, modulePath, jsDocRE)
    const jsDoc = files.get(modulePath)?.match(jsDocRE)
    return jsDoc;
  }
  catch (error) {
    throw new Error("[UnJSDoc] Error during extraction of JSDoc: "+ error)
  }
}