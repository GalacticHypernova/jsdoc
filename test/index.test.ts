import { expect, it, describe } from "vitest";
import { extractJSDoc } from "../src";

describe("unjsdoc",() => {
    it("should extract jsdoc",() => {
        const jsdoc = extractJSDoc('../playground', "hasJSDoc");
        expect(jsdoc[0]).toMatchSnapshot(`
        /**
         * Test function with JSDoc
         * @param test
         * @param test2
         * @returns {number}
         */`)
    })
    it("should return null for non-existent jsdoc", () => {
        const jsdoc = extractJSDoc("../playground", "noJSDoc")
        expect(jsdoc).toEqual(null)
    })
    it("should return null for non-existent function", () => {
        const jsdoc = extractJSDoc("../playground", "noFunction")
        expect(jsdoc).toEqual(null)
    })
})