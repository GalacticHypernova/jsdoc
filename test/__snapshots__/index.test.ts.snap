// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unjsdoc > should extract jsdoc > 
        "/**
 * Test function with JSDoc
 * @param test
 * @param test2
 * @returns {number}
 */ 1`] = `
"/**
 * Test function with JSDoc
 * @param test
 * @param test2
 * @returns {number}
 */
export function hasJSDoc"
`;

exports[`unjsdoc > should extract jsdoc > 
        /**
         * Test function with JSDoc
         * @param test
         * @param test2
         * @returns {number}
         */ 1`] = `
"/**
 * Test function with JSDoc
 * @param test
 * @param test2
 * @returns {number}
 */
export function hasJSDoc"
`;

exports[`unjsdoc > should extract jsdoc > 
        /**
 * Test function with JSDoc
 * @param test
 * @param test2
 * @returns {number}
 */ 1`] = `
"/**
 * Test function with JSDoc
 * @param test
 * @param test2
 * @returns {number}
 */
export function hasJSDoc"
`;
